### PR TITLE
riscv: linker: try to use name of register 

### PR DESCRIPTION
**Pull request by maass-hamburg**
Instead of always using the second register the SPI controller for memory mapping the jedec,spi-nor flash, use the name of the register. As a fallback, if the a register with that name doesn't exists use the second register.

To be able to add this functionality, this PR also adds DT_REG_HAS_NAME, DT_REG_ADDR_BY_NAME_OR, DT_REG_SIZE_BY_NAME_OR, DT_INST_REG_HAS_NAME, DT_INST_REG_SIZE_BY_NAME_OR and DT_INST_REG_ADDR_BY_NAME_OR. (That is also needed for https://github.com/zephyrproject-rtos/zephyr/pull/73446)

### CODE REVIEW

**Review by decsny — CHANGES_REQUESTED** [review]
need to add test cases for DT additions and update API version

File: tests/lib/devicetree/api/src/main.c
Comment (decsny) [inline tests/lib/devicetree/api/src/main.c:1]: I think it would make sense to include lines to test the "OR" fallback case

**Review by decsny — CHANGES_REQUESTED** [review]

File: tests/lib/devicetree/api/src/main.c
Comment (maass-hamburg) [reply tests/lib/devicetree/api/src/main.c:1]: These are in the lines 572, 585, 624 and 637. Do you need more than that?

**Comment by maass-hamburg** [comment]
By the way if somebody has a better name for the register address than `flash_mmap`, I'd like to hear it.
Maybe the linux dts has something already, but there I couldn't find it.

File: tests/lib/devicetree/api/src/main.c
Comment (decsny) [reply tests/lib/devicetree/api/src/main.c:1]: I see, I read it wrong

**Review by decsny — APPROVED** [review]

**Comment by maass-hamburg** [comment]
ping @galak as Assignee

**Review by 57300 — APPROVED** [review]
> By the way if somebody has a better name for the register address than `flash_mmap`, I'd like to hear it. Maybe the linux dts has something already, but there I couldn't find it.

In Zephyr, the existing SPI controllers use the name `mem` for this (while the main register block is called `control`).

Linux bindings have used a bunch of different names: `memory`, `fspi_mmap`, `qspi_mmap`, `qspi_mm`, `QuadSPI-memory`, and maybe more. So, there's not much standardization there at least.

**Pull request merged**